Megan

MeganWord opened the pull request:
End of Monday update. Menu UI is done for public tweeting only- you can view all public tweets,  create a new public tweet, and add to an existing public tweet. No users or login/private tweeting. missing two tests but I know what I need to do on them. first draft of all docstrings written.

Commit:
Revert "Userlessmenu"

Commit:
Merge pull request #2 from nss-day-cohort-13/revert-1-userlessmenu

Revert 1 userlessmenu

Commit:
most up to date info.

Commit:
can add a new chirp to a public thread.

Commit:
added some additional ux options. wrote docstrings and made sure existing tests pass.

Commit:
added missing files back

MeganWord commented:
note: only showing 6 commits here but I used mostly 'userlessmenu' branch which is merged in and deleted. it had 22 commits today.

Commit:
added two missing tests for adding tweets to the correct places.

Commit:
moved serialization into adding new and to existing chirp threads for readability.

Commit:
create a new user menu completed.

Commit:
test for adding new user passes.

Commit:
changed some menu error handling around.

Commit:
completed log-in-to-existing-user functionality I think.

Commit:
moved public or private choice to top-level for DRYness purposes.

Commit:
moved chirp index out of superfluous current_chirp object.

Commit:
edited all-chirps printer to handle private chirps. indexes will be wonky.

Commit:
wrote add-to-thread docstrings.

NewEvolution commented on Test_birdyboard.py:
```diff
@@ -0,0 +1,56 @@
+        self.public_adding_tweet = Birdyboard()
+        self.public_adding_tweet.chirps_library = {"public": [[("hey", "what"), ("no", "thank you")], [("second", "text"), ("not", "sure how this will work.")]], "private": []}
+        self.public_adding_tweet.add_new_public_chirp_thread("oh no.")
+        self.assertIn([(None, "oh no.")], self.public_adding_tweet.chirps_library["public"], )
```
Possibly a trailing extra comma here?

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,443 @@
+if __name__ == '__main__':
+    app = Birdyboard()
+    app.unlogged_in_menu_print()
+    app.unlogged_in_menu_next_step()
```
Since you're calling these both in immediate succession, and the `unlogged_in_menu_print()` method only prints, I'd probably take those prints and just tack them on to the front of the `unlogged_in_menu_next_step()` method, since that's essentially what's happening anyway.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,443 @@
+        print("Welcome. You are not logged in.\n")
+        print("1. new user\n2. log in\n3. view public chirps\n'x' to exit.\n")
+
+    def unlogged_in_menu_next_step(self):
```
Since you're calling these both in immediate succession, and the `unlogged_in_menu_print()` method only prints, I'd probably take those prints and just tack them on to the front of the `unlogged_in_menu_next_step()` method, since that's essentially what's happening anyway.

Commit:
completed add new private tweet

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,443 @@
+        print("welcome " + self.user_name + "!")
+        print("1. log out\n2. view public chirps\n3. view private chirps\n4. New public chirp thread\n5. new private chirp thread\n'x' to exit.\n")
+
+    def logged_in_menu_next_step(self):
```
Same as above here, you don't really need a whole method just to print 2 lines, the above method could be rolld into this one.

MeganWord replied:
the reason I'm running this way is because of the way I want the menus to
print. I don't want to print all the menu options every time someone types
in the wrong thing, so I pulled them out. I'll put them into "prints"
alone, but I'm not going to put them into the next steps functions.

On Tue, Aug 2, 2016 at 2:15 PM, Ryan Tanay notifications@github.com wrote:

> In birdyboard.py
> https://github.com/nss-day-cohort-13/birdyboard-MeganCFord/pull/3#discussion_r73219119
> :
> 
> > -        elif next_step == "x":  # exit.
> > -            print("goodbye.")
> > -            exit()
> > -        else:
> > -            print("command not found.")
> > -            self.unlogged_in_menu_next_step()
> >   +
> >   +# ##########################################
> >   +# ######## LOGGED IN TOP LEVEL MENU ########
> >   +# ##########################################
> >   +
> > -    def logged_in_menu_print(self):
> > -        print("welcome " + self.user_name + "!")
> > -        print("1. log out\n2. view public chirps\n3. view private chirps\n4. New public chirp thread\n5. new private chirp thread\n'x' to exit.\n")
> >   +
> > -    def logged_in_menu_next_step(self):
> 
> Same as above here, you don't really need a whole method just to print 2
> lines, the above method could be rolld into this one.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/nss-day-cohort-13/birdyboard-MeganCFord/pull/3/files/c29e8f0a6183667cfe0e6b20fd4ee0b018dec442#r73219119,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ARUAVWhD9ylZYhIDNEL5CwhttZB360KXks5qb5dugaJpZM4JaAC3
> .

Commit:
handled the make a new thread thing so it doesn't show up if the user is not logged in.

Commit:
pulled top-level menus out of functions per Ryan's suggestion.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+# ###############################
+
+        self.unlogged_in_menu = "Welcome. You are not logged in.\n1. new user\n2. log in\n3. view public chirps\n'x' to exit.\n"
+        self.logged_in_menu = "Welcome " + self.user_name + "!\n1. log out\n2. view public chirps\n3. view private chirps\n4. New public chirp thread\n5. new private chirp thread\n'x' to exit.\n"
```
This throws a hard error and crashes out both your app and your tests, as setting the value of `self.logged_in_menu` requires converting `self.user_name` to a string, and at this point in the program flow, `self.user_name` is `None` and the NoneType can't be converted to a string.

MeganWord replied:
good reason to put it in a function! ;)

On Wed, Aug 3, 2016 at 10:42 AM, Ryan Tanay notifications@github.com
wrote:

> In birdyboard.py
> https://github.com/nss-day-cohort-13/birdyboard-MeganCFord/pull/3#discussion_r73352298
> :
> 
> > +
> > -    def **init**(self):
> > -        self.users = []
> > -        self.user_name = None
> >   +
> > -        self.chirps_library = {"public": [], "private": []}
> > -        # these variables are assigned when a specific chirp thread is being viewed or added to.
> > -        self.public_or_private = ""
> > -        self.chirp_index = None
> >   +
> >   +# ###############################
> >   +# ######## MENU PRINTERS ########
> >   +# ###############################
> >   +
> > -        self.unlogged_in_menu = "Welcome. You are not logged in.\n1. new user\n2. log in\n3. view public chirps\n'x' to exit.\n"
> > -        self.logged_in_menu = "Welcome " + self.user_name + "!\n1. log out\n2. view public chirps\n3. view private chirps\n4. New public chirp thread\n5. new private chirp thread\n'x' to exit.\n"
> 
> This throws a hard error and crashes out both your app and your tests, as
> setting the value of self.logged_in_menu requires converting
> self.user_name to a string, and at this point in the program flow,
> self.user_name is None and the NoneType can't be converted to a string.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/nss-day-cohort-13/birdyboard-MeganCFord/pull/3/files/506b935ffbc8e2f3a3bd8c9c47fabec18a78913b#r73352298,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ARUAVY7VGSk-_5b-SOgUBPmUI1j8jDoAks5qcLbOgaJpZM4JaAC3
> .

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+
+# ###############################
+# ######## SERIALIZATION ########
+# ###############################
```
These methods look like a good chunk to split off into their own class.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+# ############################
+# ######## VIEW USERS ########
+# ############################
+# I am fully aware that this is a completely insecure way to handle a password system.
```
It's so insecure and such bad coding practice to store passwords that I'd strongly suggest removing password functionality entirely.  No authentication at all is better than bad authentication.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+# ########################################
+# ######## VIEW ALL CHIRPS ###############
+# ########################################
+
```
Your "View all chirps" and "View full chirp thread" methods are good candidates for rolling into their own class.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+# #############################
+# ######## ADD COMMENT ########
+# #############################
+
```
Your "Add comment" and "New chirp thread" are good candidates for rolling into their own class.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+        prints after 'view users menu'. creates input for which existing user is being selected and error handles.
+        arguments: None
+        """
+        next_step = input("'b' to go back.\n'x' to exit.\nChoose a user to log into.\n>> ")
```
Logging into a user sounds weird.  Log in _as_ makes more sense in this instance.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+        menu that appears after all chirps are printed (view_chirps). Allows the user to view a full chirp thread based on the index, or to go back or exit.
+        arguments: none
+        """
+        if user_name is not None:
```
Would this be `self.user_name`?

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,505 @@
+            if self.user_name is not None:
+                self.add_to_chirp_menu()
+            else:
+                print("command not recognized.")  # to keep an un-logged in user from pressing 1 and finding a SECRET INROAD to chirping.
```
Good catch!

Commit:
created new files for refactor.

Commit:
passing tests completed for chirp class. chirp class is not implemented in birdyboard code, which is now broken

Commit:
threader tests work except I'm working on what happens if the searcher doesn't find a match now.

Commit:
added gitignore back. not sure what happened to it.

Commit:
error handling for thread and chirp search works.

Commit:
user logic and tests completed.

Commit:
create a new user now utilizes new class code.

Commit:
log in to existing users works.

note I had to remove the user name concatenation from the logged_in_menu variable since once created with an empty user name, the variable wasn't changing to reflect the new username. printing it within users_menu_next step instead. also removed the 'find user' from user.py since I don't need to loop a dictionary. put the printer/temp list generator in there instead. This is going to happen for the chirps and threads as well.

MeganWord commented on birdyboard.py:
```diff
@@ -123,116 +123,60 @@ def create_a_user_menu(self):
                     pass
 
-    def enter_password(self, user_index):
-        """
```
@NewEvolution note I took the whole password thing out. :)

Commit:
public and private thread printer logic completed.

Commit:
threader logic works.

Commit:
edited view threads menu to handle no private chirps yet.

Commit:
view full chirp and add chirps works.

Commit:
got all tests to pass.

Commit:
cleared .txt files of test objects. final docstrings and general comment cleanup.

Commit:
added a duplicate thread-name catcher

Commit:
changed menu strings to different name from menu function.

Commit:
fixed an old menu name

Commit:
updated readme.md

NewEvolution commented on TestThreads.py:
```diff
@@ -0,0 +1,32 @@
+    def test_new_private_thread_creation(self):
+        unique_id = self.tester.generate_new_thread("another title", ("user_id", "another_user_id"))
+        tester_object = self.tester.thread_library[unique_id]
+        self.assertEqual({"title": "another title", "public": False, "allowed_users": ("user_id", "another_user_id")}, tester_object)
```
Great that you used dictionaries as opposed to another data type, your key naming makes it very clear what's going on and what all the values are for.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+
+class Birdyboard:
+    """
+    this app's functionality is a private/public message board. It constructs three subclasses which pull from three serialized .txt files: users, threads, and chirps. All items in these three files are identified by a unique UUID.
```
Technically the .txt files aren't serialized, the data is.  It'd be more correct to say "pull serialized data from three .txt files"

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+    this app's functionality is a private/public message board. It constructs three subclasses which pull from three serialized .txt files: users, threads, and chirps. All items in these three files are identified by a unique UUID.
+
+    An unlogged in user can view available logins and view (but not add to) public threads and their associated chirps.
+    A logged in user can log out, view lists of public and private threads (and add new threads), and view and add to chirps associated with threads. they can only view the private chirps they are part of.
```
Start of second sentence is uncapitalized.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+    An unlogged in user can view available logins and view (but not add to) public threads and their associated chirps.
+    A logged in user can log out, view lists of public and private threads (and add new threads), and view and add to chirps associated with threads. they can only view the private chirps they are part of.
+
+    when a new private thread is created, the UUIDs of the two associated users are saved into its information so it is only accessible through its associated user set,
```
Capitalization again.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+            print("creating a new user.")
+            self.create_a_user_menu()
+        elif next_step == "2":  # log in to a current user.
+            self.usurper.generate_users_list()
```
~~This method doesn't appear to be tested anywhere.~~

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+            self.users_menu()
+        elif next_step == "3":  # view public chirps. Does not have the option of commenting.
+            self.public_or_private = "public"
+            self.threader.generate_public_threads_list()
```
~~This method doesn't appear to be tested.~~

Commit:
docstring changes per ryan notes.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+            self.view_threads_menu()
+        elif next_step == "3":  # view private threads.
+            self.public_or_private = "private"
+            self.threader.generate_private_threads_list(self.user_id)
```
This method doesn't appear to be tested.

MeganWord replied:
I will come and talk to you about how to test a method that includes an input since that seemed impossible when i was writing my test suite.

MeganWord replied:
or how would you recommend removing the input from this method? My 'menu' birdyboard class is so long but do I need to separate it out even further/am I not separating concerns enough?

NewEvolution replied:
Yeah, never mind.   Looked at the actual method and its only output is printing.  The `generate_list` part threw me at first glance, expecting it to set `self.private_threads_list` or something.

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+        if logged_in is True:
+            thread = input(">> ")
+        else:
+            thread = input(">> ")
```
Not sure why this if/else is here, as both options do the same thing.

MeganWord replied:
good call! fixed

NewEvolution commented on birdyboard.py:
```diff
@@ -0,0 +1,421 @@
+            finally:
+                try:
+                    self.thread_id = self.threader.temp_threads[thread]
+                    self.chirper.generate_chirp_list(self.thread_id)
```
~~This method does not appear to be tested.~~

Never mind!  Saw the method, it only prints, can't really test the output.

MeganWord replied:
still not sure how I would test it since there's an 'input'- will come and talk to you about how you'd recommend splitting the method into two, I guess.

NewEvolution replied:
Doesn't need testing just because it has an input.  Both this and the user_list one are essentially menu extension methods, they don't really need testing.

NewEvolution commented on chirp.py:
```diff
@@ -0,0 +1,62 @@
+    handles the retrieval and creation of individual 'chirp' messages in birdyboard.py.
+    all chirps are saved into a dictionary with a unique UUID key, and a unique thread ID from thread.py.
+    Methods: deserialize_chirps, generate_new_chirp, serialize_chirps, generate_chirp_list.
+    """
```
I feel kind of bad nitpicking capitalization, but I'mma do it again.

Commit:
moved line in if-else, capitalized chirps.py

Commit:
capitalized thread and user.py docstrings

NewEvolution commented on user.py:
```diff
@@ -0,0 +1,63 @@
+import uuid
+
+
+class Usurper:
```
Still love this class name

NewEvolution commented:
Looking really good.  I'm über nitpicky about capitalization grammar etc. though, this could use a once-over for all the docstrings, lots of missing capitals, and having nice docstrings just makes it look polished.

Commit:
fixed bunches of grammar stuff in docstrings.

stevebrownlee commented:
@MeganCFord Just wanted to tell you what a fantastic job you did on this mastery exercise. So far, it's the most complete, well thought out code in the class.

You can continue working on it if you like, but as far as comprehension goes, you get a 👍
